precompiles/xcm/src/lib.rs

```diff
@@ -202,57 +211,69 @@ where
             return Err(revert("Assets resolution failure."));
         }
 
-        let beneficiary: MultiLocation = match beneficiary_type {
+        let beneficiary: Junction = match beneficiary_type {
             BeneficiaryType::Account32 => {
                 let recipient: [u8; 32] = input.read::<H256>()?.into();
-                X1(Junction::AccountId32 {
+                Junction::AccountId32 {
                     network: None,
                     id: recipient,
-                })
+                }
             }
             BeneficiaryType::Account20 => {
                 let recipient: H160 = input.read::<Address>()?.into();
-                X1(Junction::AccountKey20 {
+                Junction::AccountKey20 {
                     network: None,
                     key: recipient.to_fixed_bytes(),
-                })
+                }
             }
         }
         .into();
 
         let is_relay = input.read::<bool>()?;
-        let parachain_id: u32 = input.read::<U256>()?.low_u32();
-        let fee_asset_item: u32 = input.read::<U256>()?.low_u32();
 
-        if fee_asset_item as usize > assets.len() {
-            return Err(revert("Bad fee index."));
-        }
+        let parachain_id: u32 = input
+            .read::<U256>()?
+            .try_into()
+            .map_err(|_| revert("error converting parachain_id, maybe value too large"))?;
 
-        // Prepare pallet-xcm call arguments
-        let dest = if is_relay {
+        let fee_item: u32 = input
+            .read::<U256>()?
+            .try_into()
+            .map_err(|_| revert("error converting fee_index, maybe value too large"))?;
+
+        let mut destination = if is_relay {
             MultiLocation::parent()
         } else {
             X1(Junction::Parachain(parachain_id)).into_exterior(1)
         };
 
-        let assets: MultiAssets = assets
+        destination
+            .push_interior(beneficiary)
+            .map_err(|_| revert("error converting destination multilocation"))?;
+
+        let assets = assets
             .iter()
             .cloned()
             .zip(amounts.iter().cloned())
             .map(Into::into)
-            .collect::<Vec<MultiAsset>>()
-            .into();
+            .collect::<Vec<MultiAsset>>();
 
         // Build call with origin.
         let origin = Some(Runtime::AddressMapping::into_account_id(
             handle.context().caller,
         ))
         .into();
-        let call = pallet_xcm::Call::<Runtime>::reserve_withdraw_assets {
-            dest: Box::new(dest.into()),
-            beneficiary: Box::new(beneficiary.into()),
-            assets: Box::new(assets.into()),
-            fee_asset_item,
+
+        // sorting Multiassets
+        let multiassets = MultiAssets::from_sorted_and_deduplicated(assets).map_err(|_| {
+            revert("In field Assets, Provided assets either not sorted nor deduplicated")
+        })?;
```

Comment:
This could be problem if someone is just sending the same assets in array. It is highly unlikely somebody would be doing that but because it doesn't make any sense imo.  There is a potential to break so removing it.